### PR TITLE
realesrgan-ncnn-vulkan: init at 0.1.3.2

### PR DESCRIPTION
**Pull request by TilCreator**
###### Motivation for this change
I want to use Real-ESRGAN-ncnn-vulkan

###### Things done
glslang: 11.1.0 -> 11.7.0, also added a workaround
ncnn: init at 20211208
realesrgan-ncnn-vulkan: init at 0.1.3.2

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by TilCreator** [comment]
Here a full build log:
```
❯ nix-build --expr 'with import /home/tilcreator/git/nixpkgs {}; realesrgan-ncnn-vulkan'
this derivation will be built:
  /nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv
building '/nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv' on 'ssh://builder@REDACTED'...
copying path '/nix/store/ipz42ky0v2jlq3bmj38mc234yc5g3ycb-libgcc-10.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/vafyy0lx82py8403bvkaki34i5bpin2n-linux-headers-5.14' from 'https://cache.nixos.org'...
copying path '/nix/store/dq5irpn22i4wrr329bkrqlpsw298nfs2-vulkan-headers-1.2.189.1' from 'https://cache.nixos.org'...
copying path '/nix/store/lbxfixyw1yk099pjyaiy3xj5dl7kxm1g-binutils-2.35.2' from 'https://cache.nixos.org'...
copying path '/nix/store/n4fsx765038cizdq0n4wmpnxcr99nndk-expand-response-params' from 'https://cache.nixos.org'...
copying path '/nix/store/8avdf5prm7kk26af74c4ms3m7lyimfh5-giflib-5.2.1' from 'https://cache.nixos.org'...
copying path '/nix/store/q6prgn66s4achzrrnvcvyjgnm94c1bn3-glibc-2.33-56-bin' from 'https://cache.nixos.org'...
copying path '/nix/store/zgfdh4brlrjwpd7w571q9zzjvrcl30vx-libdeflate-1.8' from 'https://cache.nixos.org'...
copying path '/nix/store/5qa1r3g8s5lgvc0jfdlssq8lyb28xkwl-libjpeg-turbo-2.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/f76ccq4s20agzdz3m5fdmrs23r249fzr-libpng-apng-1.6.37' from 'https://cache.nixos.org'...
copying path '/nix/store/lrcla99rr2335m0ra8v0vrv7l68mp9fs-keyutils-1.6.3-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/a8mnmdl29vnwcrpzz2dcbya40q9zsk8f-libuv-1.42.0' from 'https://cache.nixos.org'...
copying path '/nix/store/zc25i42k172rjzm461dkz859gs8jp8kc-libssh2-1.10.0' from 'https://cache.nixos.org'...
copying path '/nix/store/qsajbzchzrpng49nfpajw3cgjcdmwrfb-libxml2-2.9.12' from 'https://cache.nixos.org'...
copying path '/nix/store/jcwyhwd086qpyd8hikmlqv1qb1kc8c2s-nghttp2-1.43.0-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/mypff9f5vvf41wrjz76dialpsj2k3wq6-libkrb5-1.18' from 'https://cache.nixos.org'...
copying path '/nix/store/yi1wv7l1p49h8gf1lapb1gi199r94k8s-protobuf-3.19.0' from 'https://cache.nixos.org'...
copying path '/nix/store/mg9wsaxhlwzqv903z62zyfplg5dz3zpq-rhash-1.4.2' from 'https://cache.nixos.org'...
copying path '/nix/store/9a417mnzmkz7nbdq3kqpa5520qlggxga-vulkan-loader-1.2.189.1' from 'https://cache.nixos.org'...
copying path '/nix/store/qc0aff3ry5n5lp3p9qwyl9fkxyprah0c-zstd-1.5.0' from 'https://cache.nixos.org'...
copying path '/nix/store/aijkgr7k9hic751amgl1kilnd2pc7gq8-libtiff-4.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/i529zgqcgh770jp1a77vi8f5yxccaadx-curl-7.79.1' from 'https://cache.nixos.org'...
copying path '/nix/store/m9m1jwnl6ll72qp366awn7ni5psx3j8c-libarchive-3.5.2-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/ismc48ybrsd6hizh43kal3gdfd4cq90g-vulkan-loader-1.2.189.1-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/alhk4bwig4mh1qqw0h7y6gkxa64kma84-glibc-2.33-56-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/h43qkabfbp44wlxbplcgwlqkdzdrm6pj-libgcc-10.3.0-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/5z49zic6nbv2vjcxzwj8li9kkk7hwj0v-libwebp-1.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/q2lj6pxd6sn75fa9pjx9x3jpgxnr6zld-cmake-3.21.2' from 'https://cache.nixos.org'...
copying path '/nix/store/iyssx2arl8bc40pjimyfxyknj06swjdx-gcc-10.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/v7w0pspwq8r2b7k2sndxq3db843z7xm5-binutils-wrapper-2.35.2' from 'https://cache.nixos.org'...
copying path '/nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/xcmlbsqabmckx42p8w18ri5zq8v2iiav-stdenv-linux' from 'https://cache.nixos.org'...
copying 3 paths...
copying path '/nix/store/hpm4cg3i642xr8cxapnc8n2g4issfyb0-ncnn-20211208' to 'ssh://builder@REDACTED'...
copying path '/nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0' to 'ssh://builder@REDACTED'...
copying path '/nix/store/vfvn84xzja689yxwzzadjq36809775y3-source' to 'ssh://builder@REDACTED'...
unpacking sources
unpacking source archive /nix/store/vfvn84xzja689yxwzzadjq36809775y3-source/src
source root is src
patching sources
configuring
fixing cmake files...
cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_SKIP_BUILD_RPATH=ON -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/share/doc/realesrgan-ncnn-vulkan -DCMAKE_INSTALL_INFODIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/include -DCMAKE_INSTALL_SBINDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin/strip -DCMAKE_RANLIB=/nix/store/lbxfixyw1yk099pjyaiy3xj5dl7kxm1g-binutils-2.35.2/bin/ranlib -DCMAKE_AR=/nix/store/lbxfixyw1yk099pjyaiy3xj5dl7kxm1g-binutils-2.35.2/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/r196j95nw050vr5rf4xa2qs590xj8ay5-Real-ESRGAN-ncnn-vulkan-0.1.3.2 -DUSE_SYSTEM_NCNN=1 -DUSE_SYSTEM_WEBP=1 -DGLSLANG_TARGET_DIR=/nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0/lib/cmake 
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Found Vulkan: /nix/store/9a417mnzmkz7nbdq3kqpa5520qlggxga-vulkan-loader-1.2.189.1/lib/libvulkan.so  
-- Found glslangValidator: /nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0/bin/glslangValidator
-- Using glslang install located at /nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0/lib/cmake
CMake Error at /nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0/lib/cmake/SPIRVTargets.cmake:102 (message):
  The following imported targets are referenced, but are missing:
  SPIRV-Tools-opt
Call Stack (most recent call first):
  CMakeLists.txt:117 (include)


-- Configuring incomplete, errors occurred!
See also "/build/src/build/CMakeFiles/CMakeOutput.log".
See also "/build/src/build/CMakeFiles/CMakeError.log".
error: build of '/nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv' on 'ssh://builder@REDACTED' failed: error: builder for '/nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv' failed with exit code 1;
              last 10 log lines:
              > CMake Error at /nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0/lib/cmake/SPIRVTargets.cmake:102 (message):
              >   The following imported targets are referenced, but are missing:
              >   SPIRV-Tools-opt
              > Call Stack (most recent call first):
              >   CMakeLists.txt:117 (include)
              >
              > 
              > -- Configuring incomplete, errors occurred!
              > See also "/build/src/build/CMakeFiles/CMakeOutput.log".
              > See also "/build/src/build/CMakeFiles/CMakeError.log".
              For full logs, run 'nix log /nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv'.
error: builder for '/nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv' failed with exit code 1;
       last 10 log lines:
       > CMake Error at /nix/store/sbz12nn3f2rrf0kcyq60qcp3jaw6z9rc-glslang-11.7.0/lib/cmake/SPIRVTargets.cmake:102 (message):
       >   The following imported targets are referenced, but are missing:
       >   SPIRV-Tools-opt
       > Call Stack (most recent call first):
       >   CMakeLists.txt:117 (include)
       >
       >
       > -- Configuring incomplete, errors occurred!
       > See also "/build/src/build/CMakeFiles/CMakeOutput.log".
       > See also "/build/src/build/CMakeFiles/CMakeError.log".
       For full logs, run 'nix log /nix/store/g3z0b41kd69gk9ssqpaiyj5v3dh9mk3f-Real-ESRGAN-ncnn-vulkan-0.1.3.2.drv'.
```

**Comment by rehno-lindeque** [comment]
This may be a naive suggestion, but given that it's asking for `SPIRV-Tools-opt`, have you tried adding [spirv-tools](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/spirv-tools/default.nix) to the build inputs?

**Comment by TilCreator** [comment]
Already tried that, sry for not writing the into the pr

File: pkgs/development/libraries/ncnn/default.nix
Comment (yrd) [inline pkgs/development/libraries/ncnn/default.nix:23]: The maintainer entry doesn't exist yet. You will need to create it in `/maintainers/maintainer-list.nix`.

File: pkgs/development/libraries/ncnn/default.nix
Comment (yrd) [inline pkgs/development/libraries/ncnn/default.nix:3]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, cmake
, opencv
, protobuf
}:
```

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (yrd) [inline pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:3]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, cmake
, vulkan-loader
, vulkan-headers
, glslang
, libgcc
, libwebp
, ncnn
}:
```

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (yrd) [inline pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:14]:
```suggestion
  };
  sourceRoot = "source/src";
```

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (yrd) [inline pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:15]:
```suggestion

  patches = [ ./cmakelists.patch ];
```

This patch seems to at least start the build process:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a234caa..2517b8e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -114,7 +114,10 @@ if(USE_SYSTEM_NCNN)
             include("${GLSLANG_TARGET_DIR}/HLSLTargets.cmake")
         endif()
         include("${GLSLANG_TARGET_DIR}/glslangTargets.cmake")
+	include("${GLSLANG_TARGET_DIR}/SPIRV-Tools/SPIRV-ToolsTarget.cmake")
+	include("${GLSLANG_TARGET_DIR}/SPIRV-Tools-opt/SPIRV-Tools-optTargets.cmake")
         include("${GLSLANG_TARGET_DIR}/SPIRVTargets.cmake")
+	include("${NCNN_TARGET_DIR}/ncnn/ncnn.cmake")
 
         if (NOT TARGET glslang OR NOT TARGET SPIRV)
             message(WARNING "glslang or SPIRV target not found! USE_SYSTEM_NCNN will be turned off.")
```

**Review by yrd — CHANGES_REQUESTED** [review]
I got the build working, or at least started. It still hangs with a bunch of errors related to the `ncnn::` namespace, but I'm not firm enough with the compiler toolkit to know what to do here...

Also so minor formatting stuff.

File: pkgs/development/libraries/ncnn/default.nix
Comment (TilCreator) [reply pkgs/development/libraries/ncnn/default.nix:23]: Ups, hoping that #147367 will get merged soon, but should change that here, too ^^

Comment (TilCreator) [reply pkgs/development/libraries/ncnn/default.nix:27]: #147367 got merged, so this is resolved

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (TilCreator) [reply pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:38]: There was more to patch, but that was a nice start, thx

**Comment by TilCreator** [comment]
@Ralith I noticed your PR #155163, should I revert the update of glslang here to make merging easier?

**Comment by Ralith** [comment]
If the update there works for you, rebasing this PR on top of it seems sensible to me. I'm also interested in feedback on using the Vulkan release cycle for glslang, vs. their own. My hope is this way we can (mostly) avoid having multiple glslangs in nixpkgs, and it seems unlikely that Khronos's release quality requirements are less stringent than glslang's own.

**Comment by TilCreator** [comment]
I will have a closer look at your PR and try to rebase it

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:18]:
```suggestion
```
There are two version numbers changing, so that isn't to useful.

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:61]:
```suggestion
    homepage = "https://github.com/xinntao/Real-ESRGAN-ncnn-vulkan";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix
Comment (TilCreator) [reply pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:18]: What would be a good solution for having 2 different version in one package, the 2 ideas I have are creating a second package (the models by themself wouldn't be of much use, but automatic updating would probably be easier) or just write the version number in the models url?

Comment (TilCreator) [reply pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:18]: Also the path for the models zip file seam random or at least not easily guessable, so that will probably stay manual work.

Comment (SuperSandro2000) [reply pkgs/tools/graphics/realesrgan-ncnn-vulkan/default.nix:18]: I think the easiest would be if we just download the latest model and update the URL.

**Comment by SuperSandro2000** [comment]
Can we squash all commits belonging to one package together? I think after that we can merge.

**Comment by TilCreator** [comment]
done

**Comment by TilCreator** [comment]
Ups... and I broke it...

**Comment by TilCreator** [comment]
Just don't look at the previous pushes ^^ 😅

**Pull request merged**
